src/System.Slices/src/System/SpanEnumerators.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System
+{
+    public partial struct Span<T>
+    {
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        /// <summary>
+        /// A struct-based enumerator, to make fast enumerations possible.
+        /// This isn't designed for direct use, instead see GetEnumerator.
+        /// </summary>
+        public struct Enumerator
+        {
```

Comment:
Once compiler sees that enumerator implements IDisposable it will be required to emit try/finally and call Dispose in there. Compiler cannot make assumptions about Dispose having no sideeffects.